Tanker cloud behaviour update

Gillogical opened the pull request:
Added code to make the tankers avoid the cloud layers at the highest densities

Commit:
Tanker cloud behaviour update

Added code to make the tankers avoid the cloud layers at the highest densities

Lugghawk commented on spawns.lua:
```diff
@@ -125,18 +125,65 @@ NorthGeorgiaFARPTransportSpawns = {
+		curRoute[i].alt = hOrbit
+		
+		-- Modify any orbit taskings
+		if #curRoute[i].task ~= nil then
```
Oof. Maybe we can make this a single check, or another method? The indentation is getting pretty smelly here. 

I think lua does do short-circuiting, so you can do:
```lua
if a ~=nil and a.b ~= nil then
...
end
```
Since if the a is nil it'll not bother checking the rest, and not cause any nil indexing.

Gillogical replied:
Wilco, I had forgotten how to write that in.

Lugghawk commented on spawns.lua:
```diff
@@ -125,18 +125,65 @@ NorthGeorgiaFARPTransportSpawns = {
+	
+	-- Calculate orbit height. Ignore for partial cloud conditions
+	local hOrbit = BladderPos.hMaxFlightAlt
+	if (((clouds.base + clouds.thickness) >= hOrbit - 305) and (clouds.density >= 7)) then
```
This check could be a separate method.

Lugghawk commented on spawns.lua:
```diff
@@ -125,18 +125,65 @@ NorthGeorgiaFARPTransportSpawns = {
+	-- Calculate orbit height. Ignore for partial cloud conditions
+	local hOrbit = BladderPos.hMaxFlightAlt
+	if (((clouds.base + clouds.thickness) >= hOrbit - 305) and (clouds.density >= 7)) then
+		hOrbit = math.min(clouds.base - 305, BladderPos.hMaxFlightAlt)
```
305 duplicated, and is a magic number. What's it represent?

Gillogical replied:
1000' (304.8m), I'll add the comments header from my original script

Lugghawk replied:
Aha. I think we can then move this into its own method on `Bladderpos` (or whatever that ends up being) that takes the clouds object and spits back the appropriate height.

Lugghawk commented on spawns.lua:
```diff
@@ -125,18 +125,65 @@ NorthGeorgiaFARPTransportSpawns = {
+
+function BladderPos.InitOnGroup(planeGroup, vSpeed)
+	local planeGroupName	= planeGroup:getName()
+	vSpeed = vSpeed or BladderPos.vSpeed
```
`local vSpeed`?

Lugghawk commented on spawns.lua:
```diff
@@ -125,18 +125,65 @@ NorthGeorgiaFARPTransportSpawns = {
+
+-- Tells a tanker (or other aircraft for that matter) to orbit between two points at a set altitude
+
+BladderPos = {}
```
What is "Bladder"? I'm reading it as the human organ but can't figure out what else it could mean.

Gillogical replied:
It's a flexible cell used to hold liquid (like the organ) - the classname I made up from Jester's "Stratobladder" designation for the KC aircraft.

Although since this specific function is pretty generic, I might rename/remove it and just name the functions a bit nicer.

Lugghawk replied:
OK cool, the liquid holder (same function as the human organ, heh) is what I was thinking of. I think renaming it could be better. Can't think of one I like yet, but I'll let you know if I do and we still need one.

Lugghawk commented on spawns.lua:
```diff
@@ -125,18 +125,65 @@ NorthGeorgiaFARPTransportSpawns = {
+BladderPos.hMaxFlightAlt	= 5486	-- meters [18 000']: Don't let the aircraft fly higher than this as Hogs won't be able to refuel. TODO: Make overrideable
+BladderPos.vSpeed			= 160		-- m/s [300kts]: Default speed of the unit
+
+function BladderPos.InitOnGroup(planeGroup, vSpeed)
```
Can we give `InitOnGroup` a better name? It might help once the above is clarified.

Commit:
Renamed to WeatherPositioning

Cleaned up based on feedback

Gillogical commented:
Renamed class to WeatherPositioning and function to avoidCloudLayer
Added clearance altitude as a constant so that it can be reused better
Attempted to clean up the function

Lugghawk commented on spawns.lua:
```diff
@@ -128,38 +128,40 @@ NorthGeorgiaFARPTransportSpawns = {
-BladderPos = {}
-BladderPos.hMaxFlightAlt	= 5486	-- meters [18 000']: Don't let the aircraft fly higher than this as Hogs won't be able to refuel. TODO: Make overrideable
-BladderPos.vSpeed			= 160		-- m/s [300kts]: Default speed of the unit
+WeatherPositioning = {}
```
Love it.

Lugghawk commented on spawns.lua:
```diff
@@ -128,38 +128,40 @@ NorthGeorgiaFARPTransportSpawns = {
-	vSpeed = vSpeed or BladderPos.vSpeed
-	
+function WeatherPositioning.getCloudFreeAltitude()
 	local clouds = env.mission.weather.clouds
```
I think the clouds object should be passed in. This is a pretty kick ass function that can be used based on whatever clouds we want. As long as we pass in something with the base/thickness attributes, we can figure out a "cloud free" altitude for it.

Lugghawk replied:
In fact, you could probably even use this function to find a de-conflicting altitude for two or more tankers.

Lugghawk replied:
If not the clouds object, then the base and thickness as two arguments.

Gillogical replied:
I'll change it to a more generic sounding deconfliction function (base, top)

Lugghawk commented on spawns.lua:
```diff
@@ -177,13 +179,13 @@ DestroyedStatics = {}
 TexacoSpawn:OnSpawnGroup(function(grp)
     scheduledSpawns[grp:getUnit(1):getName()] = {TexacoSpawn, 600}
-	BladderPos.InitOnGroup(grp, 145) -- Init against cloud base at 145m/s (280 knots)
+	WeatherPositioning.avoidCloudLayer(grp, 145) -- Init against cloud base at 145m/s (280 knots)
```
This name is great. Good pick!

Lugghawk commented on spawns.lua:
```diff
@@ -128,38 +128,40 @@ NorthGeorgiaFARPTransportSpawns = {
-							curTask.params.speed = vSpeed
-						end
-					end
+		if #curRoute[i] ~= nil and #curRoute[i].task ~= nil and #curRoute[i].task.params ~= nil and #curRoute[i].task.params.tasks ~= nil then
```
Can lua multi-line this stuff? I think a format like:
```lua
if #curRoute[i] ~= nil 
    and #curRoute[i].task ~= nil 
    and #curRoute[i].task.params ~= nil 
    and #curRoute[i].task.params.tasks ~= nil then
```

Gillogical replied:
Yep, wasn't sure if that would be neater than a single

Lugghawk commented on spawns.lua:
```diff
@@ -128,38 +128,40 @@ NorthGeorgiaFARPTransportSpawns = {
-					end
+		if #curRoute[i] ~= nil and #curRoute[i].task ~= nil and #curRoute[i].task.params ~= nil and #curRoute[i].task.params.tasks ~= nil then
+			for t = 1, #curRoute[i].task.params.tasks do
+				local curTask = curRoute[i].task.params.tasks[t]
```
If tasks is an array, can you do this as a for each loop?

```lua
for _,curTask in pairs(curRoute[i].task.params.tasks) then
...
end
```

Gillogical replied:
I was thinking of replacing it with a foreach, but that seems much neater

(edited) That's right, that is how you foreach in lua.

Lugghawk commented on spawns.lua:
```diff
@@ -128,38 +128,40 @@ NorthGeorgiaFARPTransportSpawns = {
+	return hAltitude
+end
+
+function WeatherPositioning.avoidCloudLayer(planeGroup, vSpeed)
```
Can we put a doc here to state that `vSpeed` is in meters per second instead of knots?

Commit:
Update spawns.lua

More cleanups

Note that I've also bumped the speeds as the command sets desired Ground Speed, which at 18000' would probably cause aircraft like the mirage to fall out of the sky.

I'll incorporate a TAS->IAS adjustment at a later date

Lugghawk commented on spawns.lua:
```diff
@@ -125,18 +125,76 @@ NorthGeorgiaFARPTransportSpawns = {
+		end
+	end
+	
+	local route = mist.goRoute(planeGroup, curRoute)
```
Doesn't need to be assigned if it's the last line of execution.

Commit:
Update spawns.lua
